tests/subsys/logging/log_core_additional/src/log_test.c

```diff
@@ -0,0 +1,406 @@
+
+/**
+ * @brief Log severity
+ *
+ * @details This module is registered with LOG_LEVEL_INF, LOG_LEVEL_DBG is
+ *          filtered out at compile time, only 3 message handled
+ *
+ * @addtogroup logging
+ */
+
+static void test_log_severity(void)
+{
+	log_setup(false);
+
+	backend1_cb.check_severity = true;
+	backend1_cb.exp_severity[0] = LOG_LEVEL_INF;
+	backend1_cb.exp_severity[1] = LOG_LEVEL_WRN;
+	backend1_cb.exp_severity[2] = LOG_LEVEL_ERR;
+	backend1_cb.exp_severity[3] = LOG_LEVEL_DBG;
+
+	LOG_INF("info message");
+	LOG_WRN("warning message");
+	LOG_ERR("error message");
+	LOG_DBG("debug message");
+
+	while (log_test_process(false)) {
+	}
+
+	zassert_equal(3, backend1_cb.counter,
+		      "Unexpected amount of messages received by the backend.");
+}
+
+/**
+ * @brief Customizable timestamping in log messages
+ *
+ * @details Log core permit user to register customized timestamp function
+ *
+ * @addtogroup logging
+ */
+
+static void test_log_timestamping(void)
+{
+	stamp = 0U;
+
+	log_init();
+	/* deactivate all other backend */
+	const struct log_backend *backend;
+
+	for (int i = 0; i < log_backend_count_get(); i++) {
+		backend = log_backend_get(i);
+		log_backend_deactivate(backend);
+	}
+
+	TC_PRINT("Register timestamp function\n");
+	zassert_equal(0, log_set_timestamp_func(timestamp_get, 0),
+		      "Expects successful timestamp function setting.");
+
+	memset(&backend1_cb, 0, sizeof(backend1_cb));
+	log_backend_enable(&backend1, &backend1_cb, LOG_LEVEL_DBG);
+
+	backend1_cb.check_timestamp = true;
+
+	backend1_cb.exp_timestamps[0] = 0U;
+	backend1_cb.exp_timestamps[1] = 1U;
+	backend1_cb.exp_timestamps[2] = 2U;
+
+	LOG_INF("test timestamp");
+	LOG_INF("test timestamp");
+	LOG_WRN("test timestamp");
+
+	while (log_test_process(false)) {
+	}
+
+	zassert_equal(3,
+		      backend1_cb.counter,
+		      "Unexpected amount of messages received by the backend.");
+}
+
+/**
+ * @brief Multiple logging backends
+ *
+ * @details Enable two backends in this module and enable UART backend
+ *          by CONFIG_LOG_BACKEND_UART, there are three backends at least.
+ *
+ * @addtogroup logging
+ */
+
+#define UART_BACKEND "log_backend_uart"
+static void test_multiple_backends(void)
+{
+	TC_PRINT("Test multiple backends");
+	/* enable both backend1 and backend2 */
+	log_setup(true);
+	zassert_true((log_backend_count_get() >= 2),
+		     "There is no multi backends");
+
+#ifdef CONFIG_LOG_BACKEND_UART
+	bool have_uart = false;
+	struct log_backend const *backend;
+
+	for (int i = 0; i < log_backend_count_get(); i++) {
+		backend = log_backend_get(i);
+		if (strcmp(backend->name, UART_BACKEND) == 0) {
+			have_uart = true;
+		}
+	}
+	zassert_true(have_uart, "There is no UART log backend found");
+#endif
+}
+
+/**
+ * @brief Process all logging activities using a dedicated thread
+ *
+ * @addtogroup logging
+ */
+
+#ifdef CONFIG_LOG_PROCESS_THREAD
+static void test_log_thread(void)
+{
+	TC_PRINT("Logging buffer is configured to %d bytes\n",
```

Comment:
```
if (!IS_ENABLED(CONFIG_LOG_PROCESS_THREAD)) {
   ztest_test_skip();
}
```
instead of ifdefs

Reply:
In this function, IS_ENABLED() works well, but CONFIG_LOG_BUFFER_SIZE and CONFIG_LOG_PROCESS_THREAD_STACK_SIZE always be evaluated at compile time, if they are not enabled a compile error like "CONFIG_LOG_BUFFER_SIZE undeclared" will raise.
By "ifdef", if CONFIG_LOG_PROCESS_THREAD is not enabled, this part will not be compiled.

Reply:
right, in logger source code (`log_msg.c`) i used

```
#ifndef CONFIG_LOG_BUFFER_SIZE
#define CONFIG_LOG_BUFFER_SIZE 0
#endif
```

to avoid compiler complains.